SimMuon/GEMDigitizer/src/GEMSignalModel.cc

```diff
@@ -0,0 +1,151 @@
+      resolutionX_(config.getParameter<double>("resolutionX")) {}
+
+GEMSignalModel::~GEMSignalModel() {}
+
+void GEMSignalModel::simulate(const GEMEtaPartition* roll,
+                              const edm::PSimHitContainer& simHits,
+                              CLHEP::HepRandomEngine* engine,
+                              Strips& strips_,
+                              DetectorHitMap& detectorHitMap_) {
+  bool digiMuon = false;
+  bool digiElec = false;
+  const TrapezoidalStripTopology* top(dynamic_cast<const TrapezoidalStripTopology*>(&(roll->topology())));
+  for (const auto& hit : simHits) {
+    if (std::abs(hit.particleType()) != muonPdgId && digitizeOnlyMuons_)
+      continue;
+    double elecEff = 0.;
+    double partMom = hit.pabs();
+    double checkMuonEff = CLHEP::RandFlat::shoot(engine, 0., 1.);
+    double checkElecEff = CLHEP::RandFlat::shoot(engine, 0., 1.);
+    if (std::abs(hit.particleType()) == muonPdgId && checkMuonEff < averageEfficiency_)
+      digiMuon = true;
+    if (std::abs(hit.particleType()) != muonPdgId)  //consider all non muon particles with gem efficiency to electrons
+    {
+      if (partMom <= elecMomCut1)
+        elecEff = elecEffLowCoeff * std::exp(elecEffLowParam0 * partMom * momConvFact);
+      if (partMom > elecMomCut1 && partMom < elecMomCut2)
+        elecEff = elecEffMidCoeff * log(elecEffMidParam1 * partMom * momConvFact - elecEffMidParam0) /
+                  (elecEffMidCoeff + log(elecEffMidParam1 * partMom * momConvFact - elecEffMidParam0));
+      if (partMom > elecMomCut2)
+        elecEff = 1.;
+      if (checkElecEff < elecEff)
+        digiElec = true;
+    }
+    if (!(digiMuon || digiElec))
+      continue;
+    const int bx(getSimHitBx(&hit, engine));
+    const std::vector<std::pair<int, int> >& cluster(simulateClustering(top, &hit, bx, engine));
+    for (const auto& digi : cluster) {
+      detectorHitMap_.emplace(digi, &hit);
+      strips_.emplace(digi);
+    }
+  }
+}
+
+int GEMSignalModel::getSimHitBx(const PSimHit* simhit, CLHEP::HepRandomEngine* engine) {
+  int bx = -999;
+  const LocalPoint simHitPos(simhit->localPosition());
+  const float tof(simhit->timeOfFlight());
+  // random Gaussian time correction due to electronics jitter
+  float randomJitterTime = CLHEP::RandGaussQ::shoot(engine, 0., timeJitter_);
+  const GEMDetId id(simhit->detUnitId());
+  const GEMEtaPartition* roll(geometry_->etaPartition(id));
+  if (!roll) {
+    throw cms::Exception("Geometry") << "GEMSignalModel::getSimHitBx() - GEM simhit id does not match any GEM roll id: "
+                                     << id << "\n";
+    return 999;
+  }
+  if (roll->id().region() == 0) {
+    throw cms::Exception("Geometry")
+        << "GEMSignalModel::getSimHitBx() - this GEM id is from barrel, which cannot happen: " << roll->id() << "\n";
+  }
+  const double cspeed = 299792458;  // signal propagation speed in vacuum in [m/s]
```

Comment:
it's better to get the speed of light from a central source to be consistent with other files
@civanch @ianna @cvuosalo should we still use CLHEP to get the speed of light, or is there a newer recommendation?

Reply:
I will use CLHEP/Units/GlobalPhysicalConstants.h for it.

Reply:
Yes, we are still using CLHEP/Units/GlobalPhysicalConstants.h. There is no plan at this point to replace it.